README.md

```diff
@@ -65,55 +85,152 @@ antlr4 -Dlanguage=Python3 -visitor /tests/Expr.g4
+### Base nodes
+
+A `BaseNode` subclass has fields for all rule elements and labels for all rule element labels in its corresponding grammar rule.
+Both fields and labels are available as properties on `BaseNode` instances.
+Labels take precedence over fields if the names would collide.
+
+The name of a `BaseNode` is the name of the corresponding ANTLR grammar rule, but starting with an uppercase character.
+If rule alternative labels are specified for an ANTLR rule, these are used instead of the rule name.
+
+### Transforming nodes
+
+#### Transformer
+
+The `BaseNodeTransformer` will walk over the tree from the root node to the leaf nodes.
+When visiting a node, it is possible to transform it.
+The tree is updated with transformed node before continuing the walk over the tree.
+
+To define a node transform, add a static method to the `BaseNodeTransformer` subclass passed to `process_tree`.
+
+- The name of the method you should define follows this pattern: `visit_<BaseNode>`,
+  where `<BaseNode>` should be replaced by the name of the `BaseNode` subclass to transform.
+- The method should return the transformed node.
+
+This is a simple example:
+
+```python
+class Transformer(BaseNodeTransformer):
+    @staticmethod
+    def visit_My_antlr_rule(node):
+        return node.name_of_part
+```
+
+#### Custom nodes
+
+To make it easy to return a custom node, you can define `AliasNode` subclasses.
+Normally, fields of `AliasNode`s are like symlinks to navigate an AST tree.
 
-By subclassing AstNode, we can define how Antlr4 should shape parser output to our desired AST.
-For example, `/tests/test_expr_ast.py` defines the following classes for the grammar above.
+Instances of custom nodes are created from a `BaseNode`.
+Fields and labels of the source `BaseNode` are also available on the `AliasNode`.
+If an `AliasNode` field name collides with these, it takes precedence when accessing that property.
+
+This is what a custom node looks like:
 
 ```python
-from antlr_ast import AstNode
+class NotExpr(AliasNode):
+    _fields_spec = ["expr", "op=NOT"]
+```
+
+This code defines a custom node, `NotExpr` with an `expr` and an `op` field.
+
+##### Field specs
+
+The `_fields_spec` class property is a list that defines the fields the custom node should have.
+
+This is how a field spec in this list is used when creating an custom node from a `BaseNode` (the source node):
 
-class SubExpr(AstNode):
-    _fields = ['expr->expression']
+- If a field spec does not exist on the source node, it is set to `None`
+- If multiple field specs define the same field, the first one that isn't `None` is used
+- If a field spec is just a name, it is copied from the source node
+- If a field spec is an assignment, the left side is the name of the field on the `AliasNode`
+  and the right side is the path that should be taken starting in the source node to get the node
+  that should be the value for the field on the custom node.
+  Parts of this path are separated using `.`
```

Comment:
By adding new lines, your Markdown will break the line. I know it sucks but I think here you want to write everything on the same line.

Reply:
The preview doesn't show a line break (probably because it's part of a list, and a list should be followed by an empty line).

Reply:
I think it does...

- Item 1
- Item 2 with
   stuff on
   other lines

Reply:
You can see how Github renders it here:
https://github.com/datacamp/antlr-ast/pull/16/files?short_path=04c6e90#diff-04c6e90faac2675aa89e2176d2eec7d8

Reply:
Indeed! I didn't know that, that's a good trick